### PR TITLE
refactor: Overhaul and clean up the code

### PR DESCRIPTION
**Pull request by SzczurekYT**
This pull request does several things at once, because doing them separately would be a chaos and would cause merge conflicts.
What has been done:
- The code has been cleaned up and deduplicated where possible. This includes
    - Moving or renaming methods to follow consistent style
    - `NbtTag::serialize_named` method has been deleted, because it didn't serve any meaningful purpose
    - One file was moved (the one with `Nbt` type)
    - Other misc changes I might have forgot about
- Short doc comments have been added here and there, to explain what's going on
    - Small modification was added to example code and tests to show that the nbt macro can also take variables
- Several modifications were done to better conform to rust api guidelines
    - Some traits that are recommended to be implemented were implemented (e.g. `From`)
    - Some methods now take `String` instead of taking `&str` and calling `to_owned()` due to [C-CALLER-CONTROL](https://rust-lang.github.io/api-guidelines/flexibility.html#caller-decides-where-to-copy-and-place-data-c-caller-control)

### CODE REVIEW

File: src/nbt/tag.rs [hunk 1]
```diff
@@ -74,28 +77,21 @@ impl NbtTag {
         bytes.freeze()
     }
 
-    /// Serializes as single NBT tag without name.
-    pub fn serialize_tag(&self) -> Bytes {
+    /// Writes tag id and its data into bytes
```
Comment (Norbiros) [inline src/nbt/tag.rs:80]: Before comment was used to show clear difference between serialize, and serialize_named. Now it is unnecessary.

File: src/nbt/tag.rs [hunk 2]
```diff
@@ -74,28 +77,21 @@ impl NbtTag {
...
+    pub fn serialize(&self) -> Bytes {
         let mut bytes = BytesMut::new();
         bytes.put_u8(self.id());
         bytes.put(self.serialize_raw());
         bytes.freeze()
     }
 
-    /// Serializes the NBT tag into bytes with a name and id.
-    pub fn serialize_named(&self, name: &str) -> Bytes {
```
Comment (Norbiros) [inline src/nbt/tag.rs:86]: I don't understand why you removed it. This method is used to serialize normal tag (the same way as in NBT). This can be useful. Remember, we are writing library, so why don't provide such methods?

File: src/nbt/compound.rs [hunk 3]
```diff
@@ -113,3 +112,27 @@ impl NbtCompound {
...
+    type Item = (String, NbtTag);
+
+    type IntoIter = IntoIter<String, NbtTag>;
+
+    fn into_iter(self) -> Self::IntoIter {
+        self.child_tags.into_iter()
+    }
+}
+
+impl Extend<(String, NbtTag)> for NbtCompound {
```
Comment (Norbiros) [inline src/nbt/compound.rs:134]: Change order. Things related to `(String, NbtTag)` should be close together.

File: src/nbt/compound.rs [hunk 4]
```diff
@@ -2,59 +2,58 @@ use bytes::{Buf, BufMut, Bytes, BytesMut};
...
+    pub fn new() -> NbtCompound {
+        NbtCompound {
+            child_tags: HashMap::new(),
+        }
     }
 
-    pub(crate) fn deserialize(bytes: &mut Bytes) -> NbtCompound {
-        let mut compound_tags = HashMap::new();
+    /// Reads compound data from bytes
+    pub fn deserialize_raw(bytes: &mut Bytes) -> Result<NbtCompound, Error> {
```
Comment (Norbiros) [inline src/nbt/compound.rs:24]: Why this is `raw`? You serialize NbtCompound into bytes. There are no other variants.

File: src/nbt/tag.rs [hunk 5]
```diff
@@ -25,10 +25,13 @@ pub enum NbtTag {
 }
 
 impl NbtTag {
+    /// Returns the id associated with the type of data this instance holds
```
Comment (Norbiros) [inline src/nbt/tag.rs:28]: I don't like `this instance holds` part...

File: src/nbt.rs [hunk 6]
```diff
@@ -75,3 +78,21 @@ impl Deref for Nbt {
         &self.root_tag
     }
 }
+
+impl From<NbtCompound> for Nbt {
```
Comment (Norbiros) [inline src/nbt.rs:82]: Could you add something like that but for compound, so users could eaisier modify nbt tags

**Review by Norbiros — CHANGES_REQUESTED** [review]

File: src/nbt/compound.rs (hunk 3 above)
Comment (SzczurekYT) [reply src/nbt/compound.rs:134]: I put it like this because both FromIterator and IntoIterator are about iterators

File: src/nbt/tag.rs (hunk 2 above)
Comment (SzczurekYT) [reply src/nbt/tag.rs:86]: What use do you see for this?
Note that this method can't be used to serialize data into for example a file, because the root tag needs to be a compound, while this method allows an arbitary type like a list, string, etc, so it can produce invalid nbt this way.

File: src/error.rs [hunk 7]
```diff
@@ -1,6 +1,6 @@
 use thiserror::Error;
 
-#[derive(Error, Debug)]
+#[derive(Error, Copy, Clone, Eq, PartialEq, Ord, PartialOrd, Hash, Debug)]
```
[the remark this answers is not in the document]
Comment (SzczurekYT) [reply src/error.rs:3]: Yeah, makes sense, I kinda just copy pasted all the reccomended traits where possible, without thinking much.

File: src/nbt/tag.rs (hunk 1 above)
Comment (SzczurekYT) [reply src/nbt/tag.rs:80]: Someone might be not sure what's the difference beteen serialize and serialize_raw (and deserialize respectively), hence the comment.

File: src/nbt/compound.rs (hunk 4 above)
Comment (SzczurekYT) [reply src/nbt/compound.rs:24]: I wanted to make it consitent with naming in NbtTag, which works like this
raw - only the data
not raw - type id + data

File: src/nbt/tag.rs (hunk 5 above)
Comment (SzczurekYT) [reply src/nbt/tag.rs:28]: I didn't like it too, but I wasn't able to come up with anything better.
I'll just change it to 
/// Returns the id associated with the data type

File: src/nbt/compound.rs [hunk 8]
```diff
@@ -2,59 +2,58 @@ use bytes::{Buf, BufMut, Bytes, BytesMut};
...
+#[derive(Clone, PartialEq, Debug, Default, Into)]
 pub struct NbtCompound {
     pub child_tags: HashMap<String, NbtTag>,
 }
 
 impl NbtCompound {
-    pub fn new(child_tags: HashMap<String, NbtTag>) -> NbtCompound {
-        NbtCompound { child_tags }
+    /// Creates new, empty compound
+    pub fn new() -> NbtCompound {
```
[the remark this answers is not in the document]
Comment (Norbiros) [reply src/nbt/compound.rs:17]: So why you added `Default` in derive? It does the same thing as this `new` function

File: src/nbt/compound.rs (hunk 3 above)
Comment (Norbiros) [reply src/nbt/compound.rs:134]: oh okay

File: src/nbt/tag.rs (hunk 1 above)
Comment (Norbiros) [reply src/nbt/tag.rs:80]: 1. Change order it should be `serialize`, then `serialize_raw`.
2. Describe only `serialize_raw`
3. Why to use `write` word instead of `serialize`?

File: src/nbt/compound.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (SzczurekYT) [reply src/nbt/compound.rs:17]: That's what's recommended, clippy even has a lint for that.
https://rust-lang.github.io/rust-clippy/master/index.html#/new_without_default
This also makes the type work when deriving Default for other types that hold NbtCompound.

File: src/nbt/tag.rs (hunk 1 above)
Comment (SzczurekYT) [reply src/nbt/tag.rs:80]: 1 and 2 - sure
3 - it seemed nicer when I was working on this, but serialize is good too, I'll change it

File: src/nbt/tag.rs (hunk 2 above)
Comment (Norbiros) [reply src/nbt/tag.rs:86]: In Minecraft, Mojang sometimes uses random methods, like using unnamed single tags for text components. It's better to keep these methods available, even if they seem unused, to avoid others having to create new NBT libraries lol.

File: src/nbt.rs (hunk 6 above)
Comment (SzczurekYT) [reply src/nbt.rs:82]: Good idea.

File: src/nbt/compound.rs (hunk 4 above)
Comment (Norbiros) [reply src/nbt/compound.rs:24]: Ymmm, I understand but I still think that we should change it to just `deserialize`.

Comment (SzczurekYT) [reply src/nbt/compound.rs:24]: Isn't it going to be confusing to the users of the library that methods with identical names work differently?

File: src/nbt/tag.rs (hunk 2 above)
Comment (SzczurekYT) [reply src/nbt/tag.rs:86]: > In Minecraft, Mojang sometimes uses random methods, like using unnamed single tags for text components. It's better to keep these methods available, even if they seem unused, to avoid others having to create new NBT libraries lol.

Well, we are talking about named tag, which is only used in compound. Unnamed one makes sense, because that's how type are encoded everywhere else - id and then data, and because of that we have serialize.
I don't see any good reason to keep it.

File: README.md [hunk 9]
```diff
@@ -24,24 +24,27 @@ use crab_nbt::{nbt, Nbt, NbtCompound};
...
-        "key": "value",
+        key: "value",
         "long_array": [L; 1],
         "int_array": [Int; 1],
         "list": ["a", "b", "c"],
         "nbt_inner": {
             "key": "sub value"
-        }
+        },
+        "byte_array": value,
```
[the remark this answers is not in the document]
Comment (SzczurekYT) [reply README.md:38]: There is an example of array, it's called list.
But honestly I don't like how this example turned out, I'll try to make it better

File: src/nbt/compound.rs [hunk 10]
```diff
@@ -2,59 +2,58 @@ use bytes::{Buf, BufMut, Bytes, BytesMut};
...
+        NbtCompound {
+            child_tags: HashMap::new(),
+        }
     }
 
-    pub(crate) fn deserialize(bytes: &mut Bytes) -> NbtCompound {
-        let mut compound_tags = HashMap::new();
+    /// Reads compound data from bytes
```
[the remark this answers is not in the document]
Comment (SzczurekYT) [reply src/nbt/compound.rs:23]: Why?
I think it's always better to have some docs the not.
This is pretty obvious, but still it doesn't hurt to have an addidional exploanation on what the method does.

File: src/nbt/compound.rs (hunk 4 above)
Comment (Norbiros) [reply src/nbt/compound.rs:24]: It will be confusing why compound has only `serialize_raw`, why it is raw?

File: src/nbt/compound.rs (hunk 10 above)
[the remark this answers is not in the document]
Comment (Norbiros) [reply src/nbt/compound.rs:23]: @LupusVirtute I summon you!
Tell why useless comments are bad

File: src/nbt/compound.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (Norbiros) [reply src/nbt/compound.rs:17]: So what? Why to keep `new`?

File: src/nbt/compound.rs (hunk 4 above)
Comment (SzczurekYT) [reply src/nbt/compound.rs:24]: > It will be confusing why compound has only `serialize_raw`

Ymmm, because there is no good use for having `serialize`? But we could add it if you want.
> why it is raw?

It's raw because it serializes only data.

Btw. maybe we could rename `serialize_raw` methods to `serialize_data` so the name is more clear on what it does? Because raw is a bit meaningless without a context.

File: src/nbt/compound.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (SzczurekYT) [reply src/nbt/compound.rs:17]: Because that's what people will expect, and will be confused why there is no `new` method.

File: README.md [hunk 11]
```diff
@@ -27,21 +27,33 @@ fn main() {
...
+    // The macro can also take variables to use data calculated at runtime
+    let var_as_key = "some_key".to_owned();
+    let other_key = "other_key".to_owned();
+    let value = vec![0, 1, 2];
+    let other_value = vec![3, 2, 1];
+    let nbt =  nbt!("root", {
+        var_as_key: "wohoo!",
+        "the_other_way": value,
+        other_key: other_value,
+    });
```
Comment (Norbiros) [inline README.md:48]: IMO, that is logical. We don't need to take space in readme to show such thing.

File: src/macros.rs [hunk 12]
```diff
@@ -8,15 +8,30 @@
...
+///
+/// // The macro can also take variables to use data calculated at runtime
+/// let var_as_key = "some_key".to_owned();
+/// let other_key = "other_key".to_owned();
+/// let value = vec![0, 1, 2];
+/// let other_value = vec![3, 2, 1];
+/// let nbt =  nbt!("root", {
+/// var_as_key: "wohoo!",
+/// "the_other_way": value,
+/// other_key: other_value,
```
Comment (Norbiros) [inline src/macros.rs:34]: This needs to be indented (add one tab):

File: src/nbt/compound.rs [hunk 13]
```diff
@@ -2,59 +2,58 @@ use bytes::{Buf, BufMut, Bytes, BytesMut};
 use crab_nbt::nbt::tag::NbtTag;
 use crab_nbt::nbt::utils::{get_nbt_string, END_ID};
 use derive_more::Into;
-use std::collections::HashMap;
+use std::collections::{hash_map::IntoIter, HashMap};
 
-#[derive(Debug, PartialEq, Clone, Default, Into)]
+use crate::{error::Error, Nbt};
+
+/// Compound of named nbt tags.
```
Comment (Norbiros) [inline src/nbt/compound.rs:9]: Buttery butter

File: src/nbt.rs [hunk 14]
```diff
@@ -5,16 +5,21 @@ use crab_nbt::nbt::tag::NbtTag;
 use crab_nbt::nbt::utils::*;
 use std::ops::Deref;
 
-#[derive(Debug, Clone, PartialEq)]
+pub mod compound;
+pub mod tag;
+mod utils;
+
+/// Representation of the root nbt tag.
```
Comment (Norbiros) [inline src/nbt.rs:12]: Sometimes you add dots at the end, sometimes not.

File: tests/tag.rs [hunk 15]
```diff
@@ -13,12 +13,6 @@ fn serialize_raw() {
     assert_eq!(serialized.to_vec(), 2137_i64.to_be_bytes().to_vec())
 }
 
-#[test]
-fn serialize_named() {
-    let serialized = NbtTag::Long(2137).serialize_named("hi");
-    assert_eq!(serialized.to_vec(), b"\x04\0\x02hi\0\0\0\0\0\0\x08\x59")
-}
-
```
Comment (goteusz-maszyk) [inline tests/tag.rs:21]: why did you remove this?

File: src/nbt/compound.rs (hunk 13 above)
Comment (goteusz-maszyk) [reply src/nbt/compound.rs:9]:
```suggestion
```

**Review by goteusz-maszyk — CHANGES_REQUESTED** [review]

File: tests/tag.rs (hunk 15 above)
Comment (SzczurekYT) [reply tests/tag.rs:21]: because `serialize_named` was removed so there is nothing to test

File: src/nbt/compound.rs (hunk 13 above)
Comment (SzczurekYT) [reply src/nbt/compound.rs:9]: I know, but it looks nice, when the docs are not empty.
image

File: README.md (hunk 11 above)
Comment (SzczurekYT) [reply README.md:48]: How? You are used to this kind of macro already from `serde_json` `json!` macro, but if someone never used `serde_json` and is new to this type of things he might think it's only for literals and not know variables can be used too.

File: src/macros.rs (hunk 12 above)
Comment (SzczurekYT) [reply src/macros.rs:34]: oops

File: src/nbt/compound.rs (hunk 13 above)
Comment (Norbiros) [reply src/nbt/compound.rs:9]: Lmao... Who cares about docs where they are filled with useless info

Comment (SzczurekYT) [reply src/nbt/compound.rs:9]: > Lmao... Who cares about docs where they are filled with useless info

Lmao... Who cares about useless info when the docs look good

File: src/nbt/compound.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (Norbiros) [reply src/nbt/compound.rs:17]: So remove comment.

File: src/nbt/compound.rs (hunk 10 above)
[the remark this answers is not in the document]
Comment (LupusVirtute) [reply src/nbt/compound.rs]: This is literally useless you can read it from code
we literally have all these things like methods variables and everything like they are there for something

File: src/macros.rs [hunk 16]
```diff
@@ -8,15 +8,30 @@
...
+/// let other_key = "other_key".to_owned();
+/// let value = vec![0, 1, 2];
+/// let other_value = vec![3, 2, 1];
+/// let nbt =  nbt!("root", {
+///     var_as_key: "wohoo!",
+///     "the_other_way": value,
+///     other_key: other_value,
 /// });
 /// ```
 ///
```
Comment (LupusVirtute) [inline src/macros.rs:37]: Why do we have this this shouldnt be even in history

Comment (LupusVirtute) [reply src/macros.rs:37]: Why use git when you have comments lmao

**Review by Norbiros — CHANGES_REQUESTED** [review]

File: src/nbt.rs [hunk 17]
```diff
@@ -26,16 +31,14 @@ impl Nbt {
...
-            root_tag: NbtCompound::deserialize(bytes),
+            name: get_nbt_string(bytes)?,
+            root_tag: NbtCompound::deserialize_content(bytes)?,
         })
     }
 
-    /// Reads Nbt tag, that doesn't contain the name of root compound
-    /// Used in [Network NBT](https://wiki.vg/NBT#Network_NBT_(Java_Edition))
+    /// Reads Nbt tag, that doesn't contain the name of root compound.
+    /// Used in [Network NBT](https://wiki.vg/NBT#Network_NBT_(Java_Edition)).
```
Comment (LupusVirtute) [inline src/nbt.rs:41]: Disparity in comments `NBT` and `Nbt` are we refering to Nbt object? if so clarify this

File: src/nbt.rs [hunk 18]
```diff
@@ -45,25 +48,24 @@ impl Nbt {
...
-        bytes.put(self.root_tag.serialize());
+        bytes.put(NbtTag::String(self.name.to_string()).serialize_data());
+        bytes.put(self.root_tag.serialize_content());
         bytes.freeze()
     }
 
-    /// Writes Nbt tag, without name of root compound
-    /// Used in [Network NBT](https://wiki.vg/NBT#Network_NBT_(Java_Edition))
+    /// Writes Nbt tag, without name of root compound.
+    /// Used in [Network NBT](https://wiki.vg/NBT#Network_NBT_(Java_Edition)).
```
Comment (LupusVirtute) [inline src/nbt.rs:64]: same as above

File: src/nbt.rs [hunk 19]
```diff
@@ -75,3 +77,25 @@ impl Deref for Nbt {
...
+    }
+}
+
+impl<T> AsRef<T> for Nbt
+where
+    T: ?Sized,
+    <Nbt as Deref>::Target: AsRef<T>,
+{
+    fn as_ref(&self) -> &T {
+        self.deref().as_ref()
```
Comment (LupusVirtute) [inline src/nbt.rs:93]: infinite loop?

File: src/nbt/compound.rs [hunk 20]
```diff
@@ -113,3 +108,40 @@ impl NbtCompound {
...
+}
+
+impl IntoIterator for NbtCompound {
+    type Item = (String, NbtTag);
+
+    type IntoIter = IntoIter<String, NbtTag>;
+
+    fn into_iter(self) -> Self::IntoIter {
+        self.child_tags.into_iter()
+    }
```
Comment (LupusVirtute) [inline src/nbt/compound.rs:133]: Why do you need type right here for defining type IntoIter that defines typed IntoIter that confuses

File: src/nbt.rs [hunk 21]
```diff
@@ -5,16 +5,21 @@ use crab_nbt::nbt::tag::NbtTag;
 use crab_nbt::nbt::utils::*;
 use std::ops::Deref;
 
-#[derive(Debug, Clone, PartialEq)]
+pub mod compound;
+pub mod tag;
+mod utils;
+
+/// The root nbt tag.
```
[the remark this answers is not in the document]
Comment (SzczurekYT) [reply src/nbt.rs:12]: How should I modify the doc comment then? I don't sit in your mind, so I don't know what exactly do you suggest / want.

File: src/nbt/compound.rs [hunk 22]
```diff
@@ -113,3 +108,40 @@ impl NbtCompound {
...
+}
+
+impl Extend<(String, NbtTag)> for NbtCompound {
+    fn extend<T: IntoIterator<Item = (String, NbtTag)>>(&mut self, iter: T) {
```
Comment (LupusVirtute) [inline src/nbt/compound.rs:137]: `    type Item = (String, NbtTag);`
Above you declare this but here dont use it and instead making it redeclare? like okaaaaaaaaaay

Weird things being here clean it up

File: src/nbt/tag.rs [hunk 23]
```diff
@@ -25,11 +25,21 @@ pub enum NbtTag {
...
-    pub fn serialize_raw(&self) -> Bytes {
+    pub fn serialize(&self) -> Bytes {
+        let mut bytes = BytesMut::new();
+        bytes.put_u8(self.id());
+        bytes.put(self.serialize_data());
+        bytes.freeze()
+    }
+
+    /// Serializes tag data into bytes.
+    pub fn serialize_data(&self) -> Bytes {
```
Comment (LupusVirtute) [inline src/nbt/tag.rs:42]: Okay so you just said this in comment why comment when you CAN LITERALLY NAME THE FUNCTION

File: src/nbt/tag.rs [hunk 24]
```diff
@@ -25,11 +25,21 @@ pub enum NbtTag {
 }
 
 impl NbtTag {
-    pub fn id(&self) -> u8 {
+    /// Returns the numeric id associated with the data type.
+    pub const fn id(&self) -> u8 {
```
Comment (LupusVirtute) [inline src/nbt/tag.rs:29]: why the comment when function name can function good as well i guess?

File: src/nbt/tag.rs [hunk 25]
```diff
@@ -74,29 +84,13 @@ impl NbtTag {
...
-
     pub fn deserialize(bytes: &mut Bytes) -> Result<NbtTag, Error> {
         let tag_id = bytes.get_u8();
-        Self::deserialize_raw(bytes, tag_id)
+        Self::deserialize_data(bytes, tag_id)
     }
 
-    pub fn deserialize_raw(bytes: &mut Bytes, tag_id: u8) -> Result<NbtTag, Error> {
+    /// Deserializes tag data from bytes.
+    pub fn deserialize_data(bytes: &mut Bytes, tag_id: u8) -> Result<NbtTag, Error> {
```
Comment (LupusVirtute) [inline src/nbt/tag.rs:93]: Ladies and gentelmen again why there is feature in rust like function name's when we have comments

File: tests/tag.rs [hunk 26]
```diff
@@ -2,23 +2,17 @@ use bytes::Bytes;
 use crab_nbt::{nbt, Nbt, NbtTag};
 
 #[test]
-fn serialize_raw_string() {
-    let serialized = NbtTag::String("How are you?".to_string()).serialize_raw();
+fn serialize_data_string() {
+    let serialized = NbtTag::String("How are you?".to_string()).serialize_data();
     assert_eq!(serialized.to_vec(), b"\0\x0cHow are you?")
```
Comment (LupusVirtute) [inline tests/tag.rs:7]: name `\0\x0c` in code so we know what it is?

**Review by LupusVirtute — CHANGES_REQUESTED** [review]

File: src/macros.rs (hunk 16 above)
Comment (LupusVirtute) [reply src/macros.rs:37]: apparently its documentation throw it out into seperate section like docs folder or something Or maybe use github wiki?

Comment (LupusVirtute) [reply src/macros.rs:37]: okay so my bad didnt know that it appears in rust wiki

**Comment by Norbiros** [comment]
Merged in: https://github.com/CrabCraftDev/CrabNBT/commit/ae40928a5cb4ebb648ea7c5020f510c15929ba93
